platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java

```diff
@@ -297,50 +297,106 @@ public Hash hash(
+        // of the top-level node in the chunk.
+
+        // Each chunk is processed in a separate task. Tasks have dependencies. Once all task
+        // dependencies are met, the task is scheduled for execution in the pool. Each task
+        // has N input dependencies, where N is the number of nodes at the lowest chunk rank,
+        // i.e. 2^height. Every input depenency is either set to a hash from another task,
+        // or a null value, which indicates that the input hash needs not to be recalculated,
+        // but loaded from disk. A special case of a task is leaf tasks, they are all of
+        // height 1, both input dependencies are null, but they are given a leaf instead. For
+        // these tasks, the hash is calculaded based on leaf content rather than based on input
+        // hashes.
+
+        // All tasks also have an output dependency, also a task. When a hash for the task's chunk
+        // is calculated, it is set as a input dependency of that task. Output dependency value
+        // may not be null.
+
+        // Default chunk height, from config
+        final int chunkHeight = CONFIG.virtualHasherChunkHeight();
         int firstLeafRank = Path.getRank(firstLeafPath);
         int lastLeafRank = Path.getRank(lastLeafPath);
 
         // Let the listener know we have started hashing.
         listener.onHashingStarted();
 
+        // This map contains all tasks created, but not scheduled for execution yet
         final HashMap<Long, ChunkHashTask> map = new HashMap<>();
-        ChunkHashTask resultTask = new ChunkHashTask(hashingPool, INVALID_PATH, 1);
+        // The result task. It is never executed, but used as an output dependency for
+        // the root task below. When the root task is done executing, that is it produced
+        // a root hash, this hash is set as an input dependency for this result task, where
+        // it's read and returned in the end of this method
+        ChunkHashTask resultTask = new ChunkHashTask(HASHING_POOL, INVALID_PATH, 1);
         int rootTaskHeight = Math.min(firstLeafRank, chunkHeight);
-        ChunkHashTask rootTask = new ChunkHashTask(hashingPool, ROOT_PATH, rootTaskHeight);
+        ChunkHashTask rootTask = new ChunkHashTask(HASHING_POOL, ROOT_PATH, rootTaskHeight);
         rootTask.setOut(resultTask);
         map.put(ROOT_PATH, rootTask);
 
         boolean firstLeaf = true;
         final long[] stack = new long[lastLeafRank + 1];
         Arrays.fill(stack, INVALID_PATH);
 
-        final int[] parentRankHeights = new int[256];
+        // Tasks may have different heights. The root task has a default height. If the whole
+        // virtual tree has fewer ranks than the default height, the root task will cover all
+        // the tree (almost all, see comments below about leaf task heights)
+        final int[] parentRankHeights = new int[256]; // assuming there may be no more than 256 ranks in the tree
```

Comment:
That's an interesting assumption. I hope we never get to this depth of the tree. On the other hand, I'm wondering if lowering this assumed value even further (128? 64? 32? ...16? ......8?) could bring further performance improvements to the algorithm?

Also, is this assumption enforced anywhere? Or at least asserted (preferably, with a real if/throw) anywhere? If/when it is asserted, that would be a time to move the number to a shared constant then.

Reply:
No, there are no benefits (from algorithm perspective) to have this array smaller. There are no loops like from 0 to array length.

Reply:
Sounds good.